File: typescript/src/Trie.ts
class TrieNode {
    children: { [key: string]: TrieNode };
    isEndOfWord: boolean;

    constructor() {
        this.children = {};
        this.isEndOfWord = false;
    }
}

export default class Trie {
    root: TrieNode;

    constructor() {
        this.root = new TrieNode();
    }

    insert(word: string): void {
        let current = this.root;
        for (const c of word) {
            if (!current.children[c]) {
                current.children[c] = new TrieNode();
            }
            current = current.children[c];
        }
        current.isEndOfWord = true;
    }

    find(prefix: string): string[] {
        let current = this.root;
        for (const char of prefix) {
            if (!current.children[char]) {
                return [];
            }
            current = current.children[char];
        }
        return this._findWords(current, prefix);
    }

    private _findWords(node: TrieNode, prefix: string): string[] {
        let results: string[] = [];
        if (node.isEndOfWord) {
            results.push(prefix);
        }
        for (const char in node.children) {
            results = results.concat(this._findWords(node.children[char], prefix + char));
        }
        return results;
    }

    delete(word: string): void {
        this._deleteNode(this.root, word, 0);
    }

    private _deleteNode(node: TrieNode, word: string, index: number): boolean {
        if (index === word.length) {
            if (!node.isEndOfWord) return false;
            node.isEndOfWord = false;
            return Object.keys(node.children).length === 0;
        }

        const char = word[index];
        const childNode = node.children[char];
        if (!childNode) {
            return false;
        }

        const shouldDeleteChild = this._deleteNode(childNode, word, index + 1);

        if (shouldDeleteChild) {
            delete node.children[char];
            return Object.keys(node.children).length === 0 && !node.isEndOfWord;
        }

        return false;
    }
}

File: python/src/Trie.py
''' A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. 
    There are various applications of this data structure, such as autocomplete and spellchecker.

    Implement the Trie class:

        Trie() Initializes the trie object.
        
        void insert(String word) Inserts the 'word' into the trie.
        
        boolean search(String word) Returns true if 'word' is in the trie (i.e., was inserted before), and false otherwise.
        
        boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the 'prefix', and false otherwise.'''


class Node:
    def __init__(self):
        self.children = {}
        self.isWord = False

    def __repr__(self) -> str:
        return f"Node({self.children}), isWord={self.isWord}"


class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
            if c not in curr.children:
                curr.children[c] = Node()

            curr = curr.children[c]

        curr.isWord = True

    def search(self, word: str) -> bool:
        curr = self.root
        for c in word:
            if c not in curr.children:
                return False

            curr = curr.children[c]

        return curr.isWord

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for c in prefix:
            if c not in curr.children:
                return False

            curr = curr.children[c]

        return True
